### PR TITLE
Caplin: Automatic retirement of state tables to their own snapshot files 

### PR DESCRIPTION
**Pull request by Giulio2002**

### CODE REVIEW

File: turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+	if sn.Count() != int(blocksPerFile) {
+		return fmt.Errorf("expected %d blocks, got %d", blocksPerFile, sn.Count())
+	}
+	if err := sn.Compress(); err != nil {
+		return err
+	}
+	// Generate .idx file, which is the slot => offset mapping.
+	p := &background.Progress{}
+
+	// Ugly hack to wait for fsync
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go:563]: `.Compress()` does wait for `fsync` to finish. you encountered bug? on mac?

File: cmd/capcli/cli.go [hunk 2]
```diff
@@ -1172,3 +1181,60 @@ func (c *DumpBlobsSnapshotsToStore) Run(ctx *Context) error {
...
+	}
+	snTypes := freezeblocks.MakeCaplinStateSnapshotsTypes(db)
+	stateSn := freezeblocks.NewCaplinStateSnapshots(ethconfig.BlocksFreezing{}, beaconConfig, dirs, snTypes, log.Root())
+	if err := stateSn.OpenFolder(); err != nil {
+		return err
+	}
+	r, _ := stateSn.Get(kv.BlockRoot, 999424)
+	fmt.Printf("%x\n", r)
+
+	if err := stateSn.DumpCaplinState(ctx, stateSn.BlocksAvailable(), to, c.StepSize, salt, dirs, runtime.NumCPU(), log.LvlInfo, log.Root()); err != nil {
```
[the remark this answers is not in the document]
Comment (Giulio2002) [reply cmd/capcli/cli.go:1234]: just set it to 1

File: turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go (hunk 1 above)
Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go:580]: mmh - it was there before. not sure who put it

**Review by AskAlexSharov — APPROVED** [review]

File: turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+	"github.com/erigontech/erigon/eth/ethconfig"
+)
+
+func getKvGetterForStateTable(db kv.RoDB, tableName string) KeyValueGetter {
+	return func(numId uint64) ([]byte, []byte, error) {
+		var key, value []byte
+		var err error
+		if err := db.View(context.TODO(), func(tx kv.Tx) error {
+			key = base_encoding.Encode64ToBytes4(numId)
+			value, err = tx.GetOne(tableName, base_encoding.Encode64ToBytes4(numId))
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go:57]: `value` valid only until end of transaction:
- or accept `kv.Tx` from outside to extend lifetime of `value`
- or copy it

**Review by AskAlexSharov — APPROVED** [review]

File: turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go (hunk 3 above)
Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/caplin_state_snapshots.go:57]: ii see...

**Pull request merged**